README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

WhiteSpace is a widget to create white space between widgets with the size based on the rule of 8.

<br />

## Install

```bash
flutter pub add white_space
```

<br />

## WhiteSpace Usage
| Size           	| Usage              	|
|----------------	|--------------------	|
| 2              	| `WhiteSpace.xxxs`    	|
| 4              	| `WhiteSpace.xxs`     	|
| 8              	| `WhiteSpace.xs`      	|
| 10             	| `WhiteSpace.s`       	|
| 12             	| `WhiteSpace.m`       	|
| 16             	| `WhiteSpace.l`       	|
| 24             	| `WhiteSpace.xl`      	|
| 32             	| `WhiteSpace.xxl`     	|
| 64             	| `WhiteSpace.xxxl`    	|
| Specified Size 	| `WhiteSpace.size()`  	|
| Scale of 8     	| `WhiteSpace.scale()` 	|
<br />
## SpaceSize Usage
| Size           	| Usage              	|
|----------------	|--------------------	|
| 2              	| `SpaceSize.xxxs`    	|
| 4              	| `SpaceSize.xxs`     	|
| 8              	| `SpaceSize.xs`      	|
| 10             	| `SpaceSize.s`       	|
| 12             	| `SpaceSize.m`       	|
| 16             	| `SpaceSize.l`       	|
| 24             	| `SpaceSize.xl`      	|
| 32             	| `SpaceSize.xxl`     	|
| 64             	| `SpaceSize.xxxl`    	|
<br />
## ScaleSize Usage
| Scale           	| Usage              	|
|----------------	|--------------------	|
| 0.25            	| `ScaleSize.xxxs`    	|
| 0.5            	| `ScaleSize.xxs`     	|
| 1.0            	| `ScaleSize.xs`      	|
| 1.25           	| `ScaleSize.s`       	|
| 1.5            	| `ScaleSize.m`       	|
| 2.0             	| `ScaleSize.l`       	|
| 3.0             	| `ScaleSize.xl`      	|
| 4.0             	| `ScaleSize.xxl`     	|
| 8.0             	| `ScaleSize.xxxl`    	|


CHANGELOG.md
## 0.2.0
* Add space size
* Add scale size
* Remove deprecated white space

## 0.1.0
* Improvement WhiteSpace widget
* Update docs

## 0.0.1
* create WhiteSpace widget


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:white_space/white_space.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'White Space Example',
      theme: ThemeData(primarySwatch: Colors.blue),
      home: const MyHomePage(title: 'White Space Example'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(title)),
      body: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          mainAxisAlignment: MainAxisAlignment.start,
          children: [
            const Text('The size of the white space below is 2'),
            WhiteSpace.xxxs,
            const Text('The size of the white space below is 4'),
            WhiteSpace.xxs,
            const Text('The size of the white space below is 8'),
            WhiteSpace.xs,
            const Text('The size of the white space below is 10'),
            WhiteSpace.s,
            const Text('The size of the white space below is 12'),
            WhiteSpace.m,
            const Text('The size of the white space below is 16'),
            WhiteSpace.l,
            const Text('The size of the white space below is 24'),
            WhiteSpace.xl,
            const Text('The size of the white space below is 32'),
            WhiteSpace.xxl,
            const Text('The size of the white space below is 64'),
            WhiteSpace.xxxl,
            const Text('End of Example'),
          ],
        ),
      ),
    );
  }
}


lib/src/space_size.dart
import 'base.dart';
import 'scale_size.dart';

/// Used to get the size based on the rule of 8
class SpaceSize {
  /// Get the size of 2.
  static double get xxxs => defaultSize * ScaleSize.xxxs;

  /// Get the size of 4.
  static double get xxs => defaultSize * ScaleSize.xxs;

  /// Get the size of 8.
  static double get xs => defaultSize;

  /// Get the size of 10.
  static double get s => defaultSize * ScaleSize.s;

  /// Get the size of 12.
  static double get m => defaultSize * ScaleSize.m;

  /// Get the size of 16.
  static double get l => defaultSize * ScaleSize.l;

  /// Get the size of 24.
  static double get xl => defaultSize * ScaleSize.xl;

  /// Get the size of 32.
  static double get xxl => defaultSize * ScaleSize.xxl;

  /// Get the size of 64.
  static double get xxxl => defaultSize * ScaleSize.xxxl;
}


lib/src/white_space.dart
import 'package:flutter/material.dart';

import 'base.dart';
import 'scale_size.dart';

/// A widget that returns a SizedBox widget that has the same height and width
class WhiteSpace extends StatelessWidget {
  const WhiteSpace._({Key? key, this.width, this.height}) : super(key: key);

  /// Creates a white space with the rule of 8 size.
  factory WhiteSpace.scale({
    double scale = ScaleSize.xs,
    double size = defaultSize,
  }) =>
      WhiteSpace._(width: size * scale, height: size * scale);

  /// Creates a white space with the specified size.
  factory WhiteSpace.size([double size = defaultSize]) => WhiteSpace._(
        width: size,
        height: size,
      );

  /// Creates a white space with the size of 2.
  static Widget get xxxs => WhiteSpace.scale(scale: ScaleSize.xxxs);

  /// Creates a white space with the size of 4.
  static Widget get xxs => WhiteSpace.scale(scale: ScaleSize.xxs);

  /// Creates a white space with the size of 8.
  static Widget get xs => WhiteSpace.scale();

  /// Creates a white space with the size of 10.
  static Widget get s => WhiteSpace.scale(scale: ScaleSize.s);

  /// Creates a white space with the size of 12.
  static Widget get m => WhiteSpace.scale(scale: ScaleSize.m);

  /// Creates a white space with the size of 16.
  static Widget get l => WhiteSpace.scale(scale: ScaleSize.l);

  /// Creates a white space with the size of 24.
  static Widget get xl => WhiteSpace.scale(scale: ScaleSize.xl);

  /// Creates a white space with the size of 32.
  static Widget get xxl => WhiteSpace.scale(scale: ScaleSize.xxl);

  /// Creates a white space with the size of 64.
  static Widget get xxxl => WhiteSpace.scale(scale: ScaleSize.xxxl);

  final double? width;
  final double? height;

  @override
  Widget build(BuildContext context) => SizedBox(width: width, height: height);
}
